api/client/container/utils.go

```diff
@@ -16,26 +16,29 @@ import (
 	"github.com/docker/engine-api/types/filters"
 )
 
-func waitExitOrRemoved(dockerCli *client.DockerCli, ctx context.Context, containerID string, waitRemove bool, since time.Time) (int, error) {
+func waitExitOrRemoved(dockerCli *client.DockerCli, ctx context.Context, containerID string, waitRemove bool) (chan int, error) {
 	if len(containerID) == 0 {
 		// containerID can never be empty
 		panic("Internal Error: waitExitOrRemoved needs a containerID as parameter")
 	}
 
-	var exitCode int
+	var exitCode int32
+	statusChan := make(chan int)
 	exitChan := make(chan struct{})
 	detachChan := make(chan struct{})
 	destroyChan := make(chan struct{})
+	eventsErr := make(chan error)
 
 	// Start watch events
 	eh := system.InitEventHandler()
 	eh.Handle("die", func(e events.Message) {
 		if len(e.Actor.Attributes) > 0 {
 			for k, v := range e.Actor.Attributes {
 				if k == "exitCode" {
-					var err error
-					if exitCode, err = strconv.Atoi(v); err != nil {
+					if code, err := strconv.Atoi(v); err != nil {
 						logrus.Errorf("Can't convert %q to int: %v", v, err)
+					} else {
+						atomic.StoreInt32(&exitCode, int32(code))
```

Comment:
Because channel getting data indicates the wait is finished, but indeed we still need to go on waiting for container's removal.